File: src/main/java/pe/nanamochi/banchus/services/gameplay/performance/OsuNativeCalculator.java
package pe.nanamochi.banchus.services.gameplay.performance;

import io.github.nanamochi.osu_native.wrapper.factories.DifficultyCalculatorFactory;
import io.github.nanamochi.osu_native.wrapper.factories.PerformanceCalculatorFactory;
import io.github.nanamochi.osu_native.wrapper.objects.Beatmap;
import io.github.nanamochi.osu_native.wrapper.objects.Mod;
import io.github.nanamochi.osu_native.wrapper.objects.ModsCollection;
import io.github.nanamochi.osu_native.wrapper.objects.Ruleset;
import io.github.nanamochi.osu_native.wrapper.objects.ScoreInfo;
import java.util.ArrayList;
import java.util.List;
import org.springframework.stereotype.Component;
import pe.nanamochi.banchus.entities.commons.Mods;
import pe.nanamochi.banchus.entities.db.Score;

@Component
public class OsuNativeCalculator implements PerformanceCalculator {
  @Override
  public double calculate(String beatmapPath, Score score) throws Exception {
    List<Mod> createdMods = new ArrayList<>();
    try (Beatmap beatmap = Beatmap.fromFile(beatmapPath);
        Ruleset ruleset = Ruleset.fromId(score.getMode().getValue());
        var difficultyCalculator = DifficultyCalculatorFactory.create(ruleset, beatmap);
        var performanceCalculator = PerformanceCalculatorFactory.create(ruleset);
        ModsCollection modsCollection = ModsCollection.create()) {

      for (var modInfo : Mods.fromBitmask(score.getMods())) {
        Mod mod = Mod.create(modInfo.getInitial());
        createdMods.add(mod);
        modsCollection.add(mod);
      }

      Mod classicMod = Mod.create("CL");
      createdMods.add(classicMod);
      modsCollection.add(classicMod);

      ScoreInfo scoreInfo = new ScoreInfo();
      scoreInfo.setAccuracy(score.getAccuracy() / 100.0);
      scoreInfo.setMaxCombo(score.getHighestCombo());
      scoreInfo.setCountGreat(score.getNum300s());
      scoreInfo.setCountOk(score.getNum100s());
      scoreInfo.setCountMeh(score.getNum50s());
      scoreInfo.setCountMiss(score.getNumMisses());

      var difficultyAttributes = difficultyCalculator.calculate(modsCollection);
      return performanceCalculator
          .calculate(ruleset, beatmap, modsCollection, scoreInfo, difficultyAttributes)
          .getTotal();
    } finally {
      for (Mod mod : createdMods) {
        try {
          mod.close();
        } catch (Exception e) {
          // TODO: log this or handle it in some way
        }
      }
    }
  }

  @Override
  public boolean supports(CalculatorType type) {
    return type == CalculatorType.OSU_NATIVE;
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/infrastructure/performance/OsuNativeCalculator.kt
package pe.nanamochi.banchus.infrastructure.performance

import com.github.michaelbull.result.Result
import com.github.michaelbull.result.mapError
import com.github.michaelbull.result.runCatching
import io.github.nanamochi.osu_native.wrapper.factories.DifficultyCalculatorFactory
import io.github.nanamochi.osu_native.wrapper.factories.PerformanceCalculatorFactory
import io.github.nanamochi.osu_native.wrapper.objects.Beatmap
import io.github.nanamochi.osu_native.wrapper.objects.Mod
import io.github.nanamochi.osu_native.wrapper.objects.ModsCollection
import io.github.nanamochi.osu_native.wrapper.objects.Ruleset
import io.github.nanamochi.osu_native.wrapper.objects.ScoreInfo
import org.springframework.stereotype.Component
import pe.nanamochi.banchus.score.entity.Score
import pe.nanamochi.banchus.core.enums.Mode
import pe.nanamochi.banchus.core.enums.Mods
import pe.nanamochi.banchus.core.error.CalculationFailed
import pe.nanamochi.banchus.core.error.PerformanceError

@Component
class OsuNativeCalculator : PerformanceCalculator {

    override fun calculate(beatmapPath: String, score: Score): Result<Double, PerformanceError> {
        return runCatching { Beatmap.fromFile(beatmapPath).use { it.computePerformance(score) } }
            .mapError { CalculationFailed }
    }

    override fun calculate(beatmapData: ByteArray, score: Score): Result<Double, PerformanceError> {
        return runCatching { Beatmap.fromBytes(beatmapData).use { it.computePerformance(score) } }
            .mapError { CalculationFailed }
    }

    private fun Beatmap.computePerformance(score: Score): Double {
        return Ruleset.fromId(score.mode.value.toInt()).use { ruleset ->
            val diffCalc = DifficultyCalculatorFactory.create(ruleset, this)
            val perfCalc = PerformanceCalculatorFactory.create(ruleset)

            val modNames = Mods.fromBitmask(score.mods.toUInt()).map { it.initial } + "CL"

            val mods = modNames.map { Mod.create(it) }

            mods.useEach { modsList ->
                ModsCollection.create().use { modsCollection ->
                    modsList.forEach(modsCollection::add)

                    val scoreInfo =
                        ScoreInfo().apply {
                            accuracy = score.accuracy / 100.0
                            maxCombo = score.highestCombo
                            countGreat = score.num300s
                            countOk = score.num100s
                            countMeh = score.num50s
                            countMiss = score.numMisses

                            when (score.mode) {
                                Mode.MANIA -> {
                                    countPerfect = score.numGekis
                                    countGood = score.numKatus
                                }
                                Mode.CATCH -> {
                                    countLargeTickHit = score.num100s
                                    countSmallTickHit = score.num50s
                                    countSmallTickMiss = score.numKatus
                                }
                                else -> {}
                            }
                        }

                    val diffAttrs = diffCalc.calculate(modsCollection)
                    perfCalc.calculate(ruleset, this, modsCollection, scoreInfo, diffAttrs).total
                }
            }
        }
    }

    private fun <T : AutoCloseable, R> List<T>.useEach(block: (List<T>) -> R): R {
        if (isEmpty()) return block(emptyList())

        val head = first()
        val tail = drop(1)

        return head.use { tail.useEach { processedTail -> block(listOf(head) + processedTail) } }
    }

    override fun supports(type: CalculatorType) = type == CalculatorType.OSU_NATIVE
}
